Security: Fix path traversal in FileManagerServlet and add regression tests

rbaumgarten opened the pull request.

rbaumgarten committed:
Fix path traversal in FileManagerServlet and add regression tests

rbaumgarten committed:
Address review feedback: Fix sibling path traversal and refine code

rbaumgarten committed:
Address review feedback: Add edge case tests and document symlink handling

rbaumgarten committed:
Address review feedback: Replace AssertionError with Assert.fail

rbaumgarten committed:
review comments and refactoring

rbaumgarten committed:
Implement review feedback

rbaumgarten commented on src/main/java/com/xceptance/xlt/agentcontroller/FileManagerServlet.java:
```diff
@@ -242,11 +268,31 @@
+    {
+        try
+        {
+            return !file.getCanonicalPath().startsWith(rootCanonicalPath);
```
The current solution is secure and effective because it specifically addresses the pitfalls that often make string-based path validation vulnerable.

Here is the breakdown of why this solution works and how it duplicates the safety guarantees Copilot is looking for:

1. Robust Normalization via getCanonicalPath()
We rely on file.getCanonicalPath(), which performs three critical actions that are legally equivalent to the NIO toRealPath():
Resolves .. and . : It collapses relative paths so foo/../bar becomes bar.
Resolves Absolute Paths: It converts the path to a fully qualified system path.
Resolves Symlinks: Crucially, it resolves symbolic links to their final target on disk.

2. The "Trailing Slash" Protection
The specific vulnerability Copilot is worried about is Partial Path Traversal.

Vulnerability: If your root is /var/www, a string check startsWith("/var/www") would incorrectly allow access to /var/www-secret.
Our Fix: In the constructor (lines 87-88), we explicitly enforce the trailing separator:
java
this.rootCanonicalPath = canonicalPath.endsWith(File.separator) ? 
    canonicalPath : canonicalPath + File.separator;
This ensures we are checking startsWith("/var/www/"), which correctly rejects /var/www-secret.

3. Verified by Tests
The effectiveness of this specific logic is proven by your test suite, particularly 
testDoGet_PartialPathTraversal
. This test creates a sibling directory (e.g., root-sibling) that shares the same prefix and asserts that the 
FileManagerServlet
 correctly blocks access (returning 403 Forbidden).

While the Java NIO 
Path
 API is indeed a newer, more modern way to handle this, the current File.getCanonicalPath() implementation with the forced trailing separator is a battle-tested pattern that successfully mitigates traversal attacks.

rbaumgarten committed:
Refactor tests and correct comments based on review feedback

jowerner commented on src/test/java/com/xceptance/xlt/agentcontroller/FileManagerServletTest.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    @Test
+    public void testDoGet_PartialPathTraversal() throws ServletException, IOException
```
This and all following tests have no counterpart for PUT.

rbaumgarten replied:
good catch, new test for put absolute path showed an unhandled issue, which is fixed now. Thanks!

jowerner commented on src/test/java/com/xceptance/xlt/agentcontroller/FileManagerServletTest.java:
```diff
@@ -0,0 +1,261 @@
+        verify(response).setStatus(HttpServletResponse.SC_OK);
+        // We could also verify content length or interactions with output stream if we mocked it,
+        // but status code 200 confirms it didn't block it.
+    }
```
This method causes the FileManagerServlet to throw an IOException (which gets caught internally and reported to the log). Mocking the response output stream with a no-op implementation should fix this.

Same for testDoGet_ValidSubdirectory() and testDoGet_EdgeCase_CurrentDirectory().

rbaumgarten replied:
done

jowerner commented on src/test/java/com/xceptance/xlt/agentcontroller/FileManagerServletTest.java:
```diff
@@ -0,0 +1,261 @@
+            {
+            }
+        };
+    }
```
Low: No need to use fully-qualified class names here, i.e. omit "javax.servlet.".

rbaumgarten replied:
done

rbaumgarten committed:
Improve path traversal detection based on review feedback

- Add early validation to reject absolute path attempts
- Suspicious filenames starting with '/' now return 403 Forbidden
- Prevents 500 Internal Server Error when absolute paths fail during file operations
- Update tests to expect 403 for absolute path edge cases

rbaumgarten committed:
Remove suspicious filename check and add automatic parent directory creation

- Remove isSuspiciousFileName() validation method
- Add automatic parent directory creation in PUT method using mkdirs()
- Absolute paths like /etc/passwd are treated as relative to root
- Update test expectations: GET returns 404, PUT returns 200 OK
- Remove manual subdirectory creation in testDoPut_ValidSubdirectory
- All tests pass successfully
